opensearch_py_ml/semantic search/train.py

```diff
@@ -0,0 +1,362 @@
+
+    Parameters:
+    train_examples:
+        the InputExample for the sentence transformer model training
+    model_url: str = None
+        the url to download sentence transformer model, if None, default as 'sentence-transformers/msmarco-distilbert-base-tas-b'
+    output_path: str=None
+        the path to store trained custom model. If None, default as current folder path
+    output_model_name: str=None
+        the name of the trained custom model. If None, default as 'trained_model.pt'
+    Return:
+        None
+
+    """
+
+    if output_path is None:
+        output_path = os.getcwd()
+    if output_model_name is None:
+        output_model_name = "trained_model.pt"
+    if model_url is None:
+        model_url = "sentence-transformers/msmarco-distilbert-base-tas-b"
+
+    # Load a model from HuggingFace
+    model = SentenceTransformer(model_url)
+    # prepare an output model path for later saving the pt model.
+    output_model_path = output_path + "/" + output_model_name
+    # identify if running on gpu or cpu
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    model.to(device)
+
+    corp_len = []
+    for i in range(len(train_examples)):
+        corp_len.append(len(train_examples[i].__dict__["texts"][0].split(" ")))
+
+    corp_max_len = int(np.percentile(corp_len, 90) * 1.5)
+    model.model_max_length = corp_max_len
+    model.tokenizer.model_max_length = corp_max_len
+    model.max_seq_length = corp_max_len
+
+    # num of epochs and batch will change to adapting size based on the Input Example sizes
+    num_epochs = 20
+    batch_size = 32
+    total_steps = (len(train_examples) // batch_size) * num_epochs
+    steps_size = len(train_examples) // batch_size
+
+    optimizer = torch.optim.Adam(model.parameters(), lr=2e-5)
+    scheduler = get_linear_schedule_with_warmup(
+        optimizer,
+        num_warmup_steps=max(10000, total_steps * 0.05),
+        num_training_steps=total_steps,
+    )
+    loss = []
+    init_time = time.time()
+
+    for epoch in range(num_epochs):
+        random.shuffle(train_examples)
+        print(" Training epoch " + str(epoch) + "...")
+        for j in tqdm(range(steps_size), total=steps_size):
+            batch = []
+            batch_q = []
+            for example in train_examples[j * batch_size : (j + 1) * batch_size]:
+                batch_q.append(example.texts[1])
+                batch.append(example.texts[0])
+
+            out_q = model.tokenize(batch_q)
+            for key in out_q.keys():
+                out_q[key] = out_q[key].to(device)
+            Y = model(out_q)["sentence_embedding"]
+
+            out = model.tokenize(batch)
+            for key in out.keys():
+                out[key] = out[key].to(device)
+            X = model(out)["sentence_embedding"]
+
+            XY = torch.exp(torch.matmul(X, Y.T) / (X.shape[1]) ** 0.5)
+            num = torch.diagonal(XY)
```

Comment:
if possible can you please add few comments to explain a bit more.

Reply:
will talk to Milind and see if we can add in some easy explanations later